ESQL: incomplete parsing support for query params

### Description

For a query with params like `"params": [{"x": 1}, {"y": 2}]` running in a JVM with assertions disabled results in:

```
  "error": {
    "root_cause": [
      {
        "type": "x_content_parse_exception",
        "reason": "Failed to build [params] after last required field arrived"
      }
    ],
    "type": "x_content_parse_exception",
    "reason": "[3:19] [esql/query] failed to parse field [params]",
    "caused_by": {
      "type": "x_content_parse_exception",
      "reason": "Failed to build [params] after last required field arrived",
      "caused_by": {
        "type": "array_index_out_of_bounds_exception",
        "reason": "Index 1 out of bounds for length 0"
      }
    }
  }
```

but with assertions enabled:

```
[2023-09-07T13:03:21,434][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [runTask-0] fatal error in thread [Thread-2], exiting java.lang.AssertionError: [params] must configure at least one constructor argument. If it doesn't have any it should use ObjectParser instead of ConstructingObjectParser. This is a bug in the parser declaration.
┬╗      at org.elasticsearch.xcontent@8.11.0-SNAPSHOT/org.elasticsearch.xcontent.ConstructingObjectParser$Target.finish(ConstructingObjectParser.java:543)
┬╗      at org.elasticsearch.xcontent@8.11.0-SNAPSHOT/org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166)
┬╗      at org.elasticsearch.xcontent@8.11.0-SNAPSHOT/org.elasticsearch.xcontent.ConstructingObjectParser.apply(ConstructingObjectParser.java:158)
┬╗      at org.elasticsearch.xpack.esql.action.EsqlQueryRequest.parseParams(EsqlQueryRequest.java:181)
┬╗      at org.elasticsearch.xcontent@8.11.0-SNAPSHOT/org.elasticsearch.xcontent.AbstractObjectParser.lambda$declareField$0(AbstractObjectParser.java:174)
┬╗      at org.elasticsearch.xcontent@8.11.0-SNAPSHOT/org.elasticsearch.xcontent.ObjectParser.lambda$declareField$10(ObjectParser.java:431)
┬╗      at org.elasticsearch.xcontent@8.11.0-SNAPSHOT/org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:609)
┬╗      at org.elasticsearch.xcontent@8.11.0-SNAPSHOT/org.elasticsearch.xcontent.ObjectParser.parseArray(ObjectParser.java:604)
┬╗      at org.elasticsearch.xcontent@8.11.0-SNAPSHOT/org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:643)
┬╗      at org.elasticsearch.xcontent@8.11.0-SNAPSHOT/org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315)
┬╗      at org.elasticsearch.xcontent@8.11.0-SNAPSHOT/org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:259)
┬╗      at org.elasticsearch.xcontent@8.11.0-SNAPSHOT/org.elasticsearch.xcontent.ObjectParser.apply(ObjectParser.java:382)
┬╗      at org.elasticsearch.xpack.esql.action.EsqlQueryRequest.fromXContent(EsqlQueryRequest.java:144)
┬╗      at org.elasticsearch.xpack.esql.action.RestEsqlQueryAction.prepareRequest(RestEsqlQueryAction.java:45)
```

--- a/docs/changelog/99310.yaml
+++ b/docs/changelog/99310.yaml
@@ -1,0 +1,6 @@
+pr: 99310
+summary: "ESQL: \"params\" correctly parses the values including an optional \"type\""
+area: ES|QL
+type: bug
+issues:
+ - 99294

--- a/x-pack/plugin/esql/qa/server/single-node/src/yamlRestTest/resources/rest-api-spec/test/10_basic.yml
+++ b/x-pack/plugin/esql/qa/server/single-node/src/yamlRestTest/resources/rest-api-spec/test/10_basic.yml
@@ -297,3 +297,29 @@ setup:
   - match: {columns.2.type: "keyword"}
   - length: {values: 1}
   - match: {values.0: [1, 44, "green"]}
+
+---
+"Test Mixed Input Params":
+  - do:
+      esql.query:
+        body:
+          query: 'from test | eval x = ?, y = ?, z = ?, t = ?, u = ?, v = ? | keep x, y, z, t, u, v | limit 3'
+          params: [{"value": 1, "type": "keyword"}, {"value": 2, "type": "double"}, null, true, 123, {"value": 123, "type": "long"}]
+
+  - length: {columns: 6}
+  - match: {columns.0.name: "x"}
+  - match: {columns.0.type: "keyword"}
+  - match: {columns.1.name: "y"}
+  - match: {columns.1.type: "double"}
+  - match: {columns.2.name: "z"}
+  - match: {columns.2.type: "null"}
+  - match: {columns.3.name: "t"}
+  - match: {columns.3.type: "boolean"}
+  - match: {columns.4.name: "u"}
+  - match: {columns.4.type: "integer"}
+  - match: {columns.5.name: "v"}
+  - match: {columns.5.type: "long"}
+  - length: {values: 3}
+  - match: {values.0: ["1",2.0,null,true,123,123]}
+  - match: {values.1: ["1",2.0,null,true,123,123]}
+  - match: {values.2: ["1",2.0,null,true,123,123]}

--- a/x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java
+++ b/x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/RestEsqlTestCase.java
@@ -16,6 +16,7 @@ import org.elasticsearch.client.RequestOptions;
 import org.elasticsearch.client.Response;
 import org.elasticsearch.client.ResponseException;
 import org.elasticsearch.client.WarningsHandler;
+import org.elasticsearch.common.bytes.BytesArray;
 import org.elasticsearch.common.io.Streams;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.xcontent.XContentHelper;
@@ -72,6 +73,11 @@ public class RestEsqlTestCase extends ESRestTestCase {
 
         public RequestObjectBuilder columnar(boolean columnar) throws IOException {
             builder.field("columnar", columnar);
+            return this;
+        }
+
+        public RequestObjectBuilder params(String rawParams) throws IOException {
+            builder.rawField("params", new BytesArray(rawParams).streamInput(), XContentType.JSON);
             return this;
         }
 
@@ -291,6 +297,59 @@ public class RestEsqlTestCase extends ESRestTestCase {
         var values = List.of(List.of(3, testIndexName() + "-2", 1, "id-2"), List.of(2, testIndexName() + "-1", 2, "id-1"));
 
         assertMap(result, matchesMap().entry("columns", columns).entry("values", values));
+    }
+
+    public void testErrorMessageForEmptyParams() throws IOException {
+        ResponseException re = expectThrows(
+            ResponseException.class,
+            () -> runEsql(new RequestObjectBuilder().query("row a = 1 | eval x = ?").params("[]").build())
+        );
+        assertThat(EntityUtils.toString(re.getResponse().getEntity()), containsString("Not enough actual parameters 0"));
+    }
+
+    public void testErrorMessageForInvalidParams() throws IOException {
+        ResponseException re = expectThrows(
+            ResponseException.class,
+            () -> runEsql(new RequestObjectBuilder().query("row a = 1").params("[{\"x\":\"y\"}]").build())
+        );
+        assertThat(EntityUtils.toString(re.getResponse().getEntity()), containsString("Required [value, type]"));
+    }
+
+    public void testErrorMessageForMissingTypeInParams() throws IOException {
+        ResponseException re = expectThrows(
+            ResponseException.class,
+            () -> runEsql(new RequestObjectBuilder().query("row a = 1").params("[\"x\", 123, true, {\"value\": \"y\"}]").build())
+        );
+        assertThat(EntityUtils.toString(re.getResponse().getEntity()), containsString("Required [type]"));
+    }
+
+    public void testErrorMessageForMissingValueInParams() throws IOException {
+        ResponseException re = expectThrows(
+            ResponseException.class,
+            () -> runEsql(new RequestObjectBuilder().query("row a = 1").params("[\"x\", 123, true, {\"type\": \"y\"}]").build())
+        );
+        assertThat(EntityUtils.toString(re.getResponse().getEntity()), containsString("Required [value]"));
+    }
+
+    public void testErrorMessageForInvalidTypeInParams() throws IOException {
+        ResponseException re = expectThrows(
+            ResponseException.class,
+            () -> runEsql(new RequestObjectBuilder().query("row a = 1 | eval x = ?").params("[{\"type\": \"byte\", \"value\": 5}]").build())
+        );
+        assertThat(EntityUtils.toString(re.getResponse().getEntity()), containsString("illegal data type [byte]"));
+    }
+
+    public void testErrorMessageForArrayValuesInParams() throws IOException {
+        ResponseException re = expectThrows(
+            ResponseException.class,
+            () -> runEsql(
+                new RequestObjectBuilder().query("row a = 1 | eval x = ?").params("[{\"type\": \"integer\", \"value\": [5, 6, 7]}]").build()
+            )
+        );
+        assertThat(
+            EntityUtils.toString(re.getResponse().getEntity()),
+            containsString("[params] value doesn't support values of type: START_ARRAY")
+        );
     }
 
     private static String expectedTextBody(String format, int count, @Nullable Character csvDelimiter) {

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlQueryRequest.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlQueryRequest.java
@@ -38,6 +38,8 @@ import java.util.Map;
 import java.util.function.Supplier;
 
 import static org.elasticsearch.action.ValidateActions.addValidationError;
+import static org.elasticsearch.common.xcontent.XContentParserUtils.parseFieldsValue;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
 import static org.elasticsearch.xcontent.ObjectParser.ValueType.VALUE_ARRAY;
 
 public class EsqlQueryRequest extends ActionRequest implements CompositeIndicesRequest {
@@ -47,6 +49,13 @@ public class EsqlQueryRequest extends ActionRequest implements CompositeIndicesR
         true,
         objects -> new TypedParamValue((String) objects[1], objects[0])
     );
+    private static final ParseField VALUE = new ParseField("value");
+    private static final ParseField TYPE = new ParseField("type");
+
+    static {
+        PARAM_PARSER.declareField(constructorArg(), (p, c) -> parseFieldsValue(p), VALUE, ObjectParser.ValueType.VALUE);
+        PARAM_PARSER.declareString(constructorArg(), TYPE);
+    }
 
     private static final ParseField QUERY_FIELD = new ParseField("query");
     private static final ParseField COLUMNAR_FIELD = new ParseField("columnar");
@@ -141,9 +150,7 @@ public class EsqlQueryRequest extends ActionRequest implements CompositeIndicesR
     }
 
     public static EsqlQueryRequest fromXContent(XContentParser parser) {
-        EsqlQueryRequest result = PARSER.apply(parser, null);
-        validateParams(result.params);
-        return result;
+        return PARSER.apply(parser, null);
     }
 
     private static ObjectParser<EsqlQueryRequest, Void> objectParser(Supplier<EsqlQueryRequest> supplier) {
@@ -171,7 +178,7 @@ public class EsqlQueryRequest extends ActionRequest implements CompositeIndicesR
             Object value = null;
             String type = null;
             TypedParamValue previousParam = null;
-            TypedParamValue currentParam = null;
+            TypedParamValue currentParam;
 
             while ((token = p.nextToken()) != XContentParser.Token.END_ARRAY) {
                 XContentLocation loc = p.getTokenLocation();
@@ -200,9 +207,6 @@ public class EsqlQueryRequest extends ActionRequest implements CompositeIndicesR
                         } else if (numberType == XContentParser.NumberType.LONG) {
                             value = p.longValue();
                             type = "long";
-                        } else if (numberType == XContentParser.NumberType.FLOAT) {
-                            value = p.floatValue();
-                            type = "float";
                         } else if (numberType == XContentParser.NumberType.DOUBLE) {
                             value = p.doubleValue();
                             type = "double";
@@ -242,17 +246,6 @@ public class EsqlQueryRequest extends ActionRequest implements CompositeIndicesR
     public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
         // Pass the query as the description
         return new CancellableTask(id, type, action, query, parentTaskId, headers);
-    }
-
-    protected static void validateParams(List<TypedParamValue> params) {
-        for (TypedParamValue param : params) {
-            if (param.hasExplicitType()) {
-                throw new XContentParseException(
-                    fromProto(param.tokenLocation()),
-                    "[params] must be an array where each entry is a single field (no " + "objects supported)"
-                );
-            }
-        }
     }
 
     static org.elasticsearch.xcontent.XContentLocation fromProto(ContentLocation fromProto) {

--- a/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlParser.java
+++ b/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlParser.java
@@ -135,7 +135,7 @@ public class EsqlParser {
             Token token = delegate.nextToken();
             if (token.getType() == EsqlBaseLexer.PARAM) {
                 if (param >= params.size()) {
-                    throw new ParsingException("Not enough actual parameters {} ", params.size());
+                    throw new ParsingException("Not enough actual parameters {}", params.size());
                 }
                 paramTokens.put(token, params.get(param));
                 param++;

--- a/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/action/EsqlQueryRequestTests.java
+++ b/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/action/EsqlQueryRequestTests.java
@@ -23,13 +23,16 @@ import org.elasticsearch.xcontent.NamedXContentRegistry;
 import org.elasticsearch.xcontent.XContentParser;
 import org.elasticsearch.xcontent.XContentParserConfiguration;
 import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.esql.parser.TypedParamValue;
 
 import java.io.IOException;
 import java.time.ZoneId;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.function.Supplier;
 
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
@@ -42,33 +45,18 @@ public class EsqlQueryRequestTests extends ESTestCase {
         ZoneId zoneId = randomZone();
         Locale locale = randomLocale(random());
         QueryBuilder filter = randomQueryBuilder();
-        List<Object> params = randomList(5, () -> randomBoolean() ? randomInt(100) : randomAlphaOfLength(10));
-        StringBuilder paramsString = new StringBuilder();
-        paramsString.append("[");
-        boolean first = true;
-        for (Object param : params) {
-            if (first == false) {
-                paramsString.append(", ");
-            }
-            first = false;
-            if (param instanceof String) {
-                paramsString.append("\"");
-                paramsString.append(param);
-                paramsString.append("\"");
-            } else {
-                paramsString.append(param);
-            }
-        }
-        paramsString.append("]");
+
+        List<TypedParamValue> params = randomParameters();
+        boolean hasParams = params.isEmpty() == false;
+        StringBuilder paramsString = paramsString(params, hasParams);
         String json = String.format(Locale.ROOT, """
             {
                 "query": "%s",
                 "columnar": %s,
                 "time_zone": "%s",
                 "locale": "%s",
-                "filter": %s,
-                "params": %s
-            }""", query, columnar, zoneId, locale.toLanguageTag(), filter, paramsString);
+                "filter": %s
+                %s""", query, columnar, zoneId, locale.toLanguageTag(), filter, paramsString);
 
         EsqlQueryRequest request = parseEsqlQueryRequest(json);
 
@@ -81,7 +69,7 @@ public class EsqlQueryRequestTests extends ESTestCase {
 
         assertEquals(params.size(), request.params().size());
         for (int i = 0; i < params.size(); i++) {
-            assertEquals(params.get(i), request.params().get(i).value);
+            assertEquals(params.get(i), request.params().get(i));
         }
     }
 
@@ -134,6 +122,65 @@ public class EsqlQueryRequestTests extends ESTestCase {
             .replaceAll("41770830", Long.toString(taskInfo.runningTimeNanos()))
             .trim();
         assertThat(json, equalTo(expected));
+    }
+
+    private List<TypedParamValue> randomParameters() {
+        if (randomBoolean()) {
+            return Collections.emptyList();
+        } else {
+            int len = randomIntBetween(1, 10);
+            List<TypedParamValue> arr = new ArrayList<>(len);
+            for (int i = 0; i < len; i++) {
+                boolean hasExplicitType = randomBoolean();
+                @SuppressWarnings("unchecked")
+                Supplier<TypedParamValue> supplier = randomFrom(
+                    () -> new TypedParamValue("boolean", randomBoolean(), hasExplicitType),
+                    () -> new TypedParamValue("integer", randomInt(), hasExplicitType),
+                    () -> new TypedParamValue("long", randomLong(), hasExplicitType),
+                    () -> new TypedParamValue("double", randomDouble(), hasExplicitType),
+                    () -> new TypedParamValue("null", null, hasExplicitType),
+                    () -> new TypedParamValue("keyword", randomAlphaOfLength(10), hasExplicitType)
+                );
+                arr.add(supplier.get());
+            }
+            return Collections.unmodifiableList(arr);
+        }
+    }
+
+    private StringBuilder paramsString(List<TypedParamValue> params, boolean hasParams) {
+        StringBuilder paramsString = new StringBuilder();
+        if (hasParams) {
+            paramsString.append(",\"params\":[");
+            boolean first = true;
+            for (TypedParamValue param : params) {
+                if (first == false) {
+                    paramsString.append(", ");
+                }
+                first = false;
+                if (param.hasExplicitType()) {
+                    paramsString.append("{\"type\":\"");
+                    paramsString.append(param.type);
+                    paramsString.append("\",\"value\":");
+                }
+                switch (param.type) {
+                    case "keyword" -> {
+                        paramsString.append("\"");
+                        paramsString.append(param.value);
+                        paramsString.append("\"");
+                    }
+                    case "integer", "long", "boolean", "null", "double" -> {
+                        paramsString.append(param.value);
+                    }
+                }
+                if (param.hasExplicitType()) {
+                    paramsString.append("}");
+                }
+            }
+            paramsString.append("]}");
+        } else {
+            paramsString.append("}");
+        }
+        return paramsString;
     }
 
     private static void assertParserErrorMessage(String json, String message) {

